### PR TITLE
New Storage package

### PR DESCRIPTION
**Pull request by dbanksdesign**
#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are updated
- [ ] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react/__tests__/__snapshots__/exports.ts.snap [hunk 1]
```diff
@@ -69,6 +68,8 @@ Array [
   "ToggleButtonGroup",
   "View",
   "VisuallyHidden",
+  "classNameModifier",
+  "classNameModifierByFlag",
```
Comment (calebpollman) [inline packages/react/__tests__/__snapshots__/exports.ts.snap:72]: Can we avoid adding these as public exports?

File: packages/react/__tests__/__snapshots__/exports.ts.snap [hunk 2]
```diff
@@ -69,6 +68,8 @@ Array [
   "ToggleButtonGroup",
   "View",
   "VisuallyHidden",
+  "classNameModifier",
```
Comment (reesscot) [inline packages/react/__tests__/__snapshots__/exports.ts.snap:71]: Why are these added?

File: packages/react/src/primitives/shared/index.ts [hunk 3]
```diff
@@ -1 +1,2 @@
 export { ComponentClassNames, ComponentClassObject } from './constants';
+export { classNameModifier, classNameModifierByFlag } from './utils';
```
Comment (reesscot) [inline packages/react/src/primitives/shared/index.ts:2]: Should these utils be moved to the ui package?

File: packages/react-storage/src/index.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+export * from './StorageManager';
```
Comment (calebpollman) [inline packages/react-storage/src/index.ts:1]: I'd like to move away from `export *` syntax for top level exports, way too easy for types to get exported since they aren't captured in our export snapshots

**Review by reesscot — APPROVED** [review]

File: packages/react/__tests__/__snapshots__/exports.ts.snap (hunk 1 above)
Comment (reesscot) [reply packages/react/__tests__/__snapshots__/exports.ts.snap:72]: We could make copies of them or move them to ui package

File: packages/react-storage/src/StorageManager/utils/uploadFile.ts [hunk 5]
```diff
@@ -49,7 +50,7 @@ export function uploadFile({
       resumable: false,
       progressCallback,
       contentType,
-      provider,
+      provider: provider as `AWSS3`,
```
Comment (calebpollman) [inline packages/react-storage/src/StorageManager/utils/uploadFile.ts:53]: Looks like we might be missing some commits maybe? Think @reesscot left a comment abt this in the code?

File: packages/react/__tests__/__snapshots__/exports.ts.snap (hunk 1 above)
Comment (dbanksdesign) [reply packages/react/__tests__/__snapshots__/exports.ts.snap:72]: I can move them to `/internal` ? I mean _reallllly_ these should live in `@aws-amplify/ui` so we can share them with other web frameworks

File: packages/react-storage/src/StorageManager/hooks/useUploadFiles/useUploadFiles.ts [hunk 6]
```diff
@@ -65,8 +65,8 @@ export function useUploadFiles({
           setUploadProgress({ id, progress: progressPercentage });
         };
 
-      const onError = (error) => {
-        onUploadError?.(error as string); // fixme
+      const onError = (error: string) => {
+        onUploadError?.(error); // fixme
```
Comment (calebpollman) [inline packages/react-storage/src/StorageManager/hooks/useUploadFiles/useUploadFiles.ts:69]: Can we fix this 🤣

File: packages/react/__tests__/__snapshots__/exports.ts.snap (hunk 2 above)
Comment (dbanksdesign) [reply packages/react/__tests__/__snapshots__/exports.ts.snap:71]: b/c we use them in storage manager

File: packages/react/src/primitives/shared/index.ts (hunk 3 above)
Comment (dbanksdesign) [reply packages/react/src/primitives/shared/index.ts:2]: yea probably

File: packages/react-storage/src/StorageManager/utils/uploadFile.ts (hunk 5 above)
Comment (reesscot) [reply packages/react-storage/src/StorageManager/utils/uploadFile.ts:53]: I added the one on line 44. For some reason my editor didn't show an error on this one.

File: packages/react-storage/src/index.ts (hunk 4 above)
Comment (dbanksdesign) [reply packages/react-storage/src/index.ts:1]: word

File: packages/react-storage/src/StorageManager/utils/uploadFile.ts (hunk 5 above)
Comment (dbanksdesign) [reply packages/react-storage/src/StorageManager/utils/uploadFile.ts:53]: yea  we need this for TS strict

File: packages/react-storage/src/StorageManager/StorageManager/types.ts [hunk 7]
```diff
@@ -58,7 +58,7 @@ export interface StorageManagerProps {
   /**
    * Monitor upload success
    */
-  onUploadSuccess?: (event: { key: string }) => void;
+  onUploadSuccess?: (event: { key?: string }) => void;
```
Comment (calebpollman) [inline packages/react-storage/src/StorageManager/StorageManager/types.ts:61]:
```suggestion
  onUploadSuccess?: (event: { key: string | undefined }) => void;
```

File: packages/react-storage/package.json [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  },
+  "devDependencies": {
+    "@rollup/plugin-typescript": "^8.3.1",
+    "@size-limit/preset-big-lib": "^7.0.8",
+    "@testing-library/jest-dom": "^5.14.1",
+    "@testing-library/react": "^12.0.0",
+    "@testing-library/react-hooks": "^7.0.1",
+    "@testing-library/user-event": "^13.2.1",
+    "@types/jest": "^26.0.23",
+    "@types/qrcode": "^1.4.2",
```
Comment (calebpollman) [inline packages/react-storage/package.json:67]:
```suggestion
```

**Review by reesscot — APPROVED** [review]

**Review by calebpollman — APPROVED** [review]
